azkaban-common/src/main/java/azkaban/executor/container/ContainerImplUtils.java

```diff
@@ -82,4 +90,57 @@ public static int getFlowNameHashValMapping(ExecutableFlow flow) {
     int flowNameHashVal = Math.abs(MurmurHash3.hash32x86(flowNameBytes));
     return flowNameHashVal % 100 + 1;
   }
-}
+
+  public static HashSet<String> getProxyUsersForFlow(final ProjectManager projectManager,
+      final ExecutableFlow flow) {
+    final HashSet<String> proxyUsers = new HashSet<>();
+    populateProxyUsersForFlow(flow, flow, projectManager, proxyUsers);
+    return proxyUsers;
+  }
+
+  public static void populateProxyUsersForFlow(final ExecutableFlow flow,
+      final ExecutableNode node, final ProjectManager projectManager, HashSet<String> proxyUsers) {
+    if (node instanceof ExecutableFlowBase) {
+      final ExecutableFlowBase base = (ExecutableFlowBase) node;
+      for (ExecutableNode subNode : base.getExecutableNodes()) {
+        populateProxyUsersForFlow(flow, subNode, projectManager, proxyUsers);
+      }
+    } else {
+      // If a node is disabled, we don't need to initialize its jobType container image when
+      // creating a pod.
+      if (node.getStatus() != Status.DISABLED) {
+        Project project = projectManager.getProject(flow.getProjectId());
+        Flow flowObj = project.getFlow(flow.getFlowId());
+        Props currentNodeProps = projectManager.getProperties(project, flowObj,
+            node.getId(), node.getJobSource());
+        // Get the node level property for proxy user.
+        String userToProxyFromNode = currentNodeProps.getString(USER_TO_PROXY, null);
+        // Get the node level override by user from the UI for proxy user.
+        Props currentNodeJobProps = projectManager.getJobOverrideProperty(project, flowObj,
+            node.getId(), node.getJobSource());
+        String userToProxyFromJobNode = currentNodeJobProps.getString(USER_TO_PROXY, null);
+        if (userToProxyFromJobNode != null) {
+          proxyUsers.add(userToProxyFromJobNode);
+        } else if (userToProxyFromNode != null) {
+          proxyUsers.add(userToProxyFromNode);
+        }
```

Comment:
hi, thank you for reviewing the PR. We have addressed that changing the user.to.proxy at runtime is not something we are supporting going forward as it is essential for us to prefetch credentials in our new security model ( where we need to obtain all custom credentials in the init container prior to flow start )as discussed internally. This is being done so we are intentionally unable to fetch [custom credentials](https://github.com/azkaban/azkaban/blob/8a65e94d8e9e268564cf2d864d7d7712998417c6/azkaban-hadoop-security-plugin/src/main/java/azkaban/security/AbstractHadoopSecurityManager.java#L284) after the flow starting in any way.

Reply:
I understand the problem however, even if USER_TO_PROXY changes at runtime aren't supported I believe this logic still doesn't resolve the property value correctly because upstream parent properties aren't taken into consideration. For example, let's say USER_TO_PROXY is set at flow level only and it's not specifically set by any jobs. What would be the resulting USER_TO_PROXY value for each job based on above code?

Also, It seems to me that 2 DB calls per enabled executable node may result from the `populateProxyUsersForFlow` method. In the case of complex DAGs have we analyzed how many of those do we have, how much time and number of queries will it take and what's the impact?

With regards to not allowing changing the USER_TO_PROXY value at runtime I'm a bit confused as to why are the flow parameters (or flow level runtime overwrites) checked in [KubernetesContainerizedImpl.java line ~1140 in that case?

```
if (flowParam != null && flowParam.containsKey("user.to.proxy")) {
      proxyUsersMap.add(flowParam.get("user.to.proxy"));
    }
```
In addition if USER_TO_PROXY can't be changed at runtime I believe we should add some sort of check to fail early and not the a middle of executions, that would probably result in a bad user experience.

Reply:
Sorry, I believe the objective here was unclear.
The goal here is **not** to determine the accurate proxy user required for a job. The goal is to enumerate all possible proxy users for a given flow so we can fetch appropriate custom credentials for them in the init container. 

Which is what we are doing here when we get it from the flow param -> job param --> override param. We add all of them and are not trying to evaluate which one will be used. This will be ultimately decided during execution .

Regarding the DB calls to get the node properties from the DAG, this evaluation has not been performed yet.

Reply:
My point is that the logic is not gathering all possible proxy users therefore it can lead to flow failure and very bad user experience because I don't think the reason for failure will be clearly stated in the logs